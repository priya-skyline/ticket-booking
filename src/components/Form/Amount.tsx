import React, { useState } from "react";
import { Box, TextField } from "@mui/material";

interface AmountInputProps {
  onChange: (newAmount: number) => void;
}

const AmountInput: React.FC<AmountInputProps> = ({ onChange }) => {
  const [amount, setAmount] = useState<number>(0);

  const handleAmountChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newAmount = Number(e.target.value);
    setAmount(newAmount);
    onChange(newAmount);
  };

  return (
    <Box>
      <TextField
        type="number"
        variant="outlined"
        value={amount}
        onChange={handleAmountChange}
        sx={{ mr: 1 }}
        size="small"
      />
    </Box>
  );
};

export default AmountInput;
