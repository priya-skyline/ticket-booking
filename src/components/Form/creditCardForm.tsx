import React, { useState } from "react";
import { Box, Typography, TextField, Button } from "@mui/material";

interface CreditCardFormProps {
  totalAmount: number;
}

const CreditCardForm: React.FC<CreditCardFormProps> = ({ totalAmount }) => {
  const initialFormValue = {
    total: totalAmount || 0,
    firstName: "",
    lastName: "",
    address: "",
    cardNumber: "",
    exp: "",
    cvv: "",
  };
  const [formData, setFormData] = useState(initialFormValue);

  React.useEffect(() => {
    // Update the total field in formData whenever totalAmount changes
    setFormData((prevData) => ({ ...prevData, total: totalAmount || 0 }));
  }, [totalAmount]);

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    console.log(formData);
  };

  const handleChange =
    (key: keyof typeof formData) =>
    (e: React.ChangeEvent<HTMLInputElement>) => {
      setFormData((prevData) => ({ ...prevData, [key]: e.target.value }));
    };

  return (
    <Box>
      <form onSubmit={handleSubmit}>
        {/* Total amount */}
        <Box display="flex" flexDirection="row" marginTop={1}>
          <Typography
            variant="h6"
            sx={{ marginRight: "auto" }}
            className="text-gray-600"
          >
            {`Total:`.toUpperCase()}
          </Typography>
          <Typography variant="h6" className="text-gray-600">
            ${totalAmount}
          </Typography>
        </Box>

        {/* First and Last Name */}
        <Box display="flex" flexDirection="row" marginTop={2}>
          <TextField
            placeholder="First Name"
            variant="outlined"
            sx={{ marginRight: 2 }}
            onChange={handleChange("firstName")}
            className="bg-white"
            size="small"
          />
          <TextField
            placeholder="Last Name"
            variant="outlined"
            onChange={handleChange("lastName")}
            className="bg-white"
            size="small"
          />
        </Box>

        {/* Address */}
        <TextField
          placeholder="Address"
          variant="outlined"
          fullWidth
          margin="normal"
          onChange={handleChange("address")}
          className="bg-white"
          size="small"
        />

        {/* Payment Details */}
        <Typography
          marginTop={2}
          marginBottom={1}
          className="text-gray-700 text-sm"
        >
          Payment Details
        </Typography>

        {/* Credit Card Number */}
        <TextField
          placeholder="Credit Card Number"
          variant="outlined"
          type="number"
          fullWidth
          sx={{ marginRight: 2 }}
          onChange={handleChange("cardNumber")}
          className="bg-white"
          size="small"
        />
        <Box display="flex" flexDirection="row" marginTop={1}>
          <TextField
            placeholder="Exp"
            variant="outlined"
            sx={{ marginRight: 2 }}
            onChange={handleChange("exp")}
            className="bg-white"
            size="small"
          />
          <TextField
            type="number"
            placeholder="CVV"
            variant="outlined"
            onChange={handleChange("cvv")}
            className="bg-white"
            size="small"
            inputProps={{
              maxLength: 3,
            }}
          />
        </Box>

        {/* Submit Button */}
        <Box mt={2}>
          <Button type="submit" variant="contained" color="primary" fullWidth>
            Get Tickets
          </Button>
        </Box>
      </form>
    </Box>
  );
};

export default CreditCardForm;
