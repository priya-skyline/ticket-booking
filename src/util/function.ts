export function convertCentsToDollars(cents: number) {
  return cents / 100;
}

export const UnixTimestampToDate = (timestamp: number) => {
  // const timestamp = 1686675030000; // Replace this with your timestamp

  const date = new Date(timestamp);
  const options: Intl.DateTimeFormatOptions = {
    weekday: "long",
    month: "long",
    day: "numeric",
  };

  return new Intl.DateTimeFormat("en-US", options).format(date);
};
