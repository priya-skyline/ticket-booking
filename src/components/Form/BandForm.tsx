import React, { useState } from "react";
import { Grid, Box, Typography, Divider } from "@mui/material";
import AmountInput from "./Amount";
import { convertCentsToDollars } from "../../util/function";

interface TicketType {
  name: string;
  description: string;
  cost: number;
}

interface Band {
  name: string;
  ticketTypes: TicketType[];
}

interface BandFormProps {
  band: Band;
  onTotalAmountChange: (totalAmount: number) => void;
}

const BandForm: React.FC<BandFormProps> = ({ band, onTotalAmountChange }) => {
  const [amounts, setAmounts] = useState<number[]>(
    Array(band.ticketTypes.length).fill(0)
  );

  const handleAmountChange = (index: number, newAmount: number) => {
    setAmounts((prevAmounts) => {
      const newAmounts = [...prevAmounts];
      newAmounts[index] = newAmount;
      return newAmounts;
    });
  };

  const totalAmount = band.ticketTypes.reduce((total, ticket, index) => {
    return total + amounts[index] * ticket.cost;
  }, 0);

  // Notify the parent component about the total amount change
  React.useEffect(() => {
    onTotalAmountChange(totalAmount);
  }, [totalAmount, onTotalAmountChange]);

  return (
    <Box>
      <Box>
        {band.ticketTypes.map((ticket, index) => (
          <React.Fragment key={ticket.name}>
            <Grid container spacing={2} mb={2}>
              <Grid item xs={9}>
                <Box>
                  <Typography pt={2} variant="h6" className="text-gray-500">
                    {ticket.name.toUpperCase()}
                  </Typography>
                  <Typography pt={2} variant="body2" className="text-gray-400">
                    {ticket.description}
                  </Typography>
                  <Typography pt={2} className="text-gray-500">
                    ${convertCentsToDollars(ticket.cost)}
                  </Typography>
                </Box>
              </Grid>

              <Grid item xs={3}>
                <Box pt={2}>
                  <AmountInput
                    onChange={(newAmount) =>
                      handleAmountChange(index, newAmount)
                    }
                  />
                </Box>
              </Grid>
            </Grid>
            <Divider sx={{ my: 2 }} />
          </React.Fragment>
        ))}
      </Box>
    </Box>
  );
};

export default BandForm;
