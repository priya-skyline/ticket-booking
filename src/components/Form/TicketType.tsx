import React from "react";
import BandForm from "./BandForm";
import { Box } from "@mui/material";
import kpopBand from "../../api/band-json/kpop-band.json";

interface TicketTypeProps {
  onTotalAmountChange: (totalAmount: number) => void;
}

export const TicketType: React.FC<TicketTypeProps> = ({
  onTotalAmountChange,
}) => {
  return (
    <Box>
      <BandForm band={kpopBand} onTotalAmountChange={onTotalAmountChange} />
    </Box>
  );
};
