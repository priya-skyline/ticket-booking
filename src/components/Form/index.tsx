import React from "react";
import { Paper, Grid, Box, Typography, useMediaQuery } from "@mui/material";
import CalendarMonthIcon from "@mui/icons-material/CalendarMonth";
import PlaceIcon from "@mui/icons-material/Place";
import parse from "html-react-parser";

import { TicketType } from "./TicketType";
import {
  UnixTimestampToDate,
  convertCentsToDollars,
} from "../../util/function";

import kpopBand from "../../api/band-json/kpop-band.json";
import CreditCardForm from "./creditCardForm";

const Form: React.FC = () => {
  const isMobile = useMediaQuery("(max-width: 768px)");
  const [totalAmount, setTotalAmount] = React.useState<number>(0);
  const handleTotalAmountChange = (newTotalAmount: number) => {
    setTotalAmount(newTotalAmount);
  };

  return (
    <Box width={isMobile ? "100%" : "50%"} margin="0 auto">
      <Paper elevation={3} className="p-12 justify-center items-center">
        <Box>
          <Typography
            variant="h4"
            gutterBottom
            className="text-gray-600"
            style={{ fontWeight: 600 }}
          >
            {kpopBand.name}
          </Typography>
        </Box>
        <Box display="flex" alignItems="center">
          <CalendarMonthIcon className="text-gray-500" />
          <Typography variant="body1" className="text-gray-400">
            {UnixTimestampToDate(kpopBand.date)}
          </Typography>
        </Box>
        <Box display="flex" alignItems="center" className="pt-3">
          <PlaceIcon className="text-gray-500" />
          <Typography variant="body1" className="text-gray-400">
            {kpopBand.location}
          </Typography>
        </Box>

        <Box className="mt-6">
          <Grid container spacing={1}>
            <Grid item xs={12} sm={12} md={12} lg={5} p={2}>
              <Box>
                <Box className="h-56 overflow-hidden flex justify-center items-center rounded-md">
                  <img
                    alt="Profile"
                    src={kpopBand.imgUrl}
                    className="object-cover"
                  />
                </Box>
                <Box>
                  <Typography
                    variant="body1"
                    mt={2}
                    gutterBottom
                    className="text-gray-600"
                  >
                    {parse(kpopBand.description_blurb)}
                  </Typography>
                </Box>
              </Box>
            </Grid>
            <Grid item xs={12} sm={12} md={12} lg={7} p={2}>
              <Box
                className="pt-6 pl-10 pr-10 pb-6 rounded-md"
                style={{ background: "#f7f8fa" }}
              >
                <Typography
                  variant="h5"
                  className="text-gray-700"
                  style={{ fontWeight: 600 }}
                >
                  Select Tickets
                </Typography>
                {/* render each ticket type */}
                <TicketType onTotalAmountChange={handleTotalAmountChange} />

                <CreditCardForm
                  totalAmount={convertCentsToDollars(totalAmount)}
                />
              </Box>
              <Box></Box>
            </Grid>
          </Grid>
        </Box>
      </Paper>
    </Box>
  );
};

export default Form;
