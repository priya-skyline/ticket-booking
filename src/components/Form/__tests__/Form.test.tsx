import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "@testing-library/react";
import Form from "..";
import { convertCentsToDollars, getCurrentTime } from "../../../util/function";

describe("<Form />", () => {
  test("renders Form component", () => {
    render(<Form />);
    // Test if specific text elements are present
    expect(screen.getByText("The Mac & Cheeses")).toBeInTheDocument();
    expect(screen.getByText(getCurrentTime())).toBeInTheDocument();

    // Test if an image is present
    expect(screen.getByAltText("Profile Image")).toBeInTheDocument();
  });

  test("getCurrentTime function returns correct value", () => {
    const toLocaleDateStringMock = jest
      .spyOn(global.Date.prototype, "toLocaleDateString")
      .mockReturnValue("Tuesday, November 10");

    const result = getCurrentTime();

    // Assertions
    expect(result).toBe("Tuesday, November 10");

    // Restore the original function to avoid affecting other tests
    toLocaleDateStringMock.mockRestore();
  });

  test("convertCentsToDollars function converts cents to dollars correctly", () => {
    // Test when 100 cents are passed, it should return 2.5 dollars
    expect(convertCentsToDollars(100)).toBe(1);

    // Test when 250 cents are passed, it should return 2.5 dollars
    expect(convertCentsToDollars(250)).toBe(2.5);

    // Test for an edge case (0 cents)
    expect(convertCentsToDollars(0)).toBe(0);
  });

  test("matches snapshot", () => {
    const { asFragment } = render(<Form />);

    expect(asFragment()).toMatchSnapshot();
  });
});
