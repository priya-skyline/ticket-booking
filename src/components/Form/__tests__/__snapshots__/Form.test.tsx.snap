// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-hk5t7d"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 form-paper css-3su884-MuiPaper-root"
    >
      <div
        class="MuiBox-root css-0"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
        >
          The Mac & Cheeses
        </h4>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-ofrcno-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <div
            class="MuiBox-root css-70qvj9"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CalendarMonthIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 4h-1V2h-2v2H8V2H6v2H5c-1.11 0-1.99.9-1.99 2L3 20c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 16H5V10h14v10zM9 14H7v-2h2v2zm4 0h-2v-2h2v2zm4 0h-2v-2h2v2zm-8 4H7v-2h2v2zm4 0h-2v-2h2v2zm4 0h-2v-2h2v2z"
              />
            </svg>
            <p
              class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
            >
              Friday, November 10
            </p>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-70qvj9"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="PlaceIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
            />
          </svg>
          <p
            class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
          >
            Surat, India
          </p>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 left-form css-k3f0g4-MuiGrid-root"
        >
          1st Partxx
          <div
            class="MuiBox-root css-0"
          >
            <img
              alt="Profile Image"
              src="https://via.placeholder.com/600/51aa97"
              style="width: 90%;"
            />
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
            </p>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7 right-form css-uqwprf-MuiGrid-root"
        >
          2nd part
          <div
            class="bg-blue-500 text-white p-4 MuiBox-root css-0"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              Select Tickets
            </h5>
            <p
              class="MuiTypography-root MuiTypography-body1 ticket-heading css-ahj2mt-MuiTypography-root"
            >
              General Admission
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              $1
            </p>
          </div>
          <div
            class="MuiBox-root css-0"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
